### ISSUE TITLE
[stable-cadence] Relax default entitlement mapping for dictionary/array elements

### ISSUE THREAD

**Issue by SupunS** [onflow/cadence#2953]
### Issue to be solved

Currently, there is no way for dictionaries/arrays to define the entitlement mapping for their children. So the default behavior is to have no entitlements for references when a child is accessed through a reference to the parent.

However, this can be restrictive sometimes, if one gets a mutable reference to a dictionary `auth(Mutate) {String: {String: AnyStruct}}` and wants to mutate the inner dictionary, then this is a bit of a hassle. In order to mutate the inner one, they would need to either:
- Get a dereference copy of the child, update it, and insert it again.
- Remove the child, update it, and insert it again.

Both of these are not only expensive but also involve a lot of boilerplate code.

e.g: https://discord.com/channels/613813861610684416/1176642288986894366/1176949623638732912

### Suggested Solution

An alternative is to give `Identity` entitlement mapping for children. So children would get the same entitlement as the parent.

**Comment by SupunS** [comment]
cc: @joshuahannan

**Comment by turbolent** [comment]
@SupunS Could you please add the concerns to the description or as a comment?

**Comment by turbolent** [comment]
Example:

```cadence
let xs = {"a": {"b": "c"}}
let x: ??? = xs[1]
```

**Comment by dsainati1** [comment]
My concerns here are primarily about safety; I don't think it is necessarily the case that the entitlements I have to an outer collection should automatically propagate to the inner collections. E.g. just because I have `Mutate` permissions to an array of NFTs, allowing me to insert or remove NFTs from the array, does not mean I should necessarily have permissions to modify the NFTs themselves.

**Comment by dsainati1** [comment]
Yeah I think this should be possible. If it isn't currently, that seems like the right way to solve this.

**Comment by SupunS** [comment]
The problem here is for dictionaries and arrays, where there is no way for a user to define custom getters/setters or entitlement mappings.
